### PR TITLE
fix #2442 assertj migration

### LINKED ISSUE #2442: Migrate JUnit 4 assertions to assert4J
This is the last step that will also permit removing the JUnit 4 jar from the test classpath, following the migration of tests to JUnit 5.

### PR DESCRIPTION
**Pull request by ericbottard**
This migrates junit4 Assertions to assert4j and gets rid of Junit4 jar.

I propose that someone reviews this PR, and then I'd be happy to do the merge to 3.4 with the live review from someone else (maybe @aneveu ?), as we did for the recent test PR.

### CODE REVIEW

**Review by aneveu — APPROVED** [review]

File: reactor-core/src/test/java/reactor/core/publisher/scenarios/FluxTests.java [hunk 1]
```diff
@@ -562,7 +561,7 @@ public void analyticsTest() throws Exception {
 		}
 		source.onComplete();
 
-		Assert.assertTrue(result.block(Duration.ofSeconds(5)) >= avgTime * 0.6);
+		assertThat(result.block(Duration.ofSeconds(5))).isGreaterThan((long)(avgTime * 0.6));
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/scenarios/FluxTests.java:564]: should be `isGreaterThanOrEqualTo`

File: reactor-core/src/test/java/reactor/core/publisher/scenarios/FluxTests.java [hunk 2]
```diff
@@ -1099,7 +1097,7 @@ public void shouldCorrectlyDispatchBatchedTimeout() throws InterruptedException
 			throw new RuntimeException(latch.getCount()+"");
 		}
 		else {
-			assertEquals("Must have correct latch number : " + latch.getCount(), latch.getCount(), 0);
+			assertThat(0).as("Must have correct latch number : " + latch.getCount()).isEqualTo(latch.getCount());
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/scenarios/FluxTests.java:1100]: invert expected and asserted values

File: reactor-core/src/test/java/reactor/core/scheduler/SingleSchedulerTest.java [hunk 3]
```diff
@@ -113,14 +114,14 @@ public void lotsOfTasks() throws Exception {
...
-    	        Assert.assertTrue(cdl.await(5, TimeUnit.SECONDS));
+    	        assertThat(cdl.await(5, TimeUnit.SECONDS)).isTrue();
 
     	        System.gc();
     	        Thread.sleep(200);
 
                 long after = ManagementFactory.getMemoryMXBean().getHeapMemoryUsage().getUsed();
 
-    	        Assert.assertTrue(String.format("%,d -> %,d", before, after), before + 20_000_000 > after);
+    	        assertThat(before + 20_000_000 > after).as(format("%,d -> %,d", before, after)).isTrue();
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/scheduler/SingleSchedulerTest.java:124]: could assert `before + 20_000_000` `isGreaterThan(after)`

**Review by simonbasle — CHANGES_REQUESTED** [review]

File: reactor-core/src/test/java/reactor/core/publisher/EmitterProcessorTest.java [hunk 4]
```diff
@@ -217,7 +217,7 @@ public void onComplete() {
 		latch.await(8, TimeUnit.SECONDS);
 
 		long count = latch.getCount();
-		assertThat(latch.getCount() == 0).as("Count > 0 : " + count + "  , Running on " + DEFAULT_POOL_SIZE + " CPU").isTrue();
+		assertThat(latch.getCount()).as("Count > 0 : %d, Running on %s CPU").isEqualTo(0);
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/EmitterProcessorTest.java:220]: missing the values for the placeholders

File: reactor-core/src/test/java/reactor/core/publisher/FluxBufferPredicateTest.java [hunk 5]
```diff
@@ -1186,7 +1186,7 @@ public void discardRaceWithOnNext_bufferWhileWithNoMatch() {
 			Assertions.assertThat(value2).as("trigger value not discarded").hasValue(-2);
 
 			Assertions.assertThat(received.get() + (value1.get() + 1))
-					.as("received " + received.get() + ", val1 state " + value1.get())
+					.as("received %d, val1 state %d")
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/FluxBufferPredicateTest.java:1189]: missing the values for the placeholders

File: reactor-core/src/test/java/reactor/core/publisher/FluxPeekFuseableTest.java [hunk 6]
```diff
@@ -493,7 +493,7 @@ public void syncFusionAvailable() {
 		    .subscribe(ts);
 
 		Subscription s = ts.upstream();
-		assertThat(s).as("Non-fuseable upstream: " + s).isInstanceOf(QueueSubscription.class);
+		assertThat(s).as("Non-fuseable upstream").isInstanceOf(QueueSubscription.class);
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/FluxPeekFuseableTest.java:496]: wondering about the removal of `s` here. can we make sure the message displays the toString of  `s` anyway?

File: reactor-core/src/test/java/reactor/core/publisher/FluxSubscribeOnValueTest.java [hunk 7]
```diff
@@ -54,7 +54,7 @@ public void testSubscribeOnValueFusion() {
 		int minExec = 2;
 
 		for (Integer counted : execs.values()) {
-			assertThat(counted).as("Thread used less than " + minExec + " " + "times").isGreaterThanOrEqualTo(minExec);
+			assertThat(counted).as("Thread used less than %d times").isGreaterThanOrEqualTo(minExec);
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/FluxSubscribeOnValueTest.java:57]: missing the values for the placeholders

File: reactor-core/src/test/java/reactor/core/scheduler/BoundedElasticSchedulerTest.java [hunk 8]
```diff
@@ -484,7 +484,7 @@ public void lifoEvictionNoThreadRegrowth() throws InterruptedException {
 					.withFailMessage("worker thread count after shutdown is not zero. threads: %s", Thread.getAllStackTraces().keySet())
 					.isNotPositive();
 			assertThat(dumpThreadNames())
-					.as("current evictor " + newEvictor + " shutdown")
+					.as("current evictor %s shutdown")
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/scheduler/BoundedElasticSchedulerTest.java:487]: missing the values for the placeholders

**Review by simonbasle — CHANGES_REQUESTED** [review]

File: reactor-core/src/test/java/reactor/core/publisher/FluxPeekTest.java [hunk 9]
```diff
@@ -716,7 +716,7 @@ public void syncFusionAvailable() {
 		    .subscribe(ts);
 
 		Subscription s = ts.upstream();
-		assertThat(s instanceof QueueSubscription).as("Non-fuseable upstream: " + s).isTrue();
+		assertThat(s).as("Non-fuseable upstream").isInstanceOf(QueueSubscription.class);
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/FluxPeekTest.java:719]: same as in previous file with the removal of `s` from message

**Review by simonbasle — CHANGES_REQUESTED** [review]

File: reactor-core/src/test/java/reactor/core/publisher/EmitterProcessorTest.java [hunk 10]
```diff
@@ -163,9 +163,8 @@ public void onComplete() {
 		latch.await(8, TimeUnit.SECONDS);
 
 		long count = latch.getCount();
-		org.junit.Assert.assertTrue("Count > 0 : " + count + " (" + list + ")  , Running on " +
-						Schedulers.DEFAULT_POOL_SIZE + " CPU",
-				latch.getCount() == 0);
+		assertThat(latch.getCount()).as("Count > 0 : " + count + " (" + list + ").isTrue()  , Running on " +
```
[the remark this answers is not in the document]
Comment (simonbasle) [reply reactor-core/src/test/java/reactor/core/publisher/EmitterProcessorTest.java:166]: still not fixed

File: reactor-core/src/test/java/reactor/core/publisher/FluxCancelOnTest.java [hunk 11]
```diff
@@ -52,7 +51,7 @@ public void cancelOnDedicatedScheduler() throws Exception {
 		    .cancel();
 
 		latch.await();
-		Assert.assertNull(threadHash.get());
+		assertThat(threadHash.get()).isNull();
```
[the remark this answers is not in the document]
Comment (simonbasle) [reply reactor-core/src/test/java/reactor/core/publisher/FluxCancelOnTest.java:54]: still not fixed

File: reactor-core/src/test/java/reactor/core/publisher/FluxPeekFuseableTest.java (hunk 6 above)
Comment (ericbottard) [reply reactor-core/src/test/java/reactor/core/publisher/FluxPeekFuseableTest.java:496]: It does, but there is actually another problem: assertJ tries to be smart when printing `s` and tries to call `iterator()` on it. And so we get
```
Although QueueSubscription extends Queue it is purely internal and only guarantees support for poll/clear/size/isEmpty. Instances shouldn't be used/exposed as Queue outside of Reactor operators.
```
which is a design problem in the first place IMHO (Liskov substitution principle violation)

Comment (simonbasle) [reply reactor-core/src/test/java/reactor/core/publisher/FluxPeekFuseableTest.java:496]: `withFailMessage` should fix that I think

**Review by simonbasle — APPROVED** [review]

**Comment by simonbasle** [comment]
(forward merge is ongoing, a lot of conflicts to review)

**Comment by simonbasle** [comment]
this will be closed by a push of a squashed commit to `3.3.x`, followed up by a push of the large merge commit on `master` (reviewed off-band due to the difficulty of reviewing merge commits in general, on github in particular)

**Pull request merged**